qualtran/bloqs/state_preparation/state_preparation_via_rotation.py

```diff
@@ -0,0 +1,443 @@
+            ]
+        else:
+            angles = [
+                np.angle(c) - offset for c, offset in zip(self.state_coefficients, offset_angles)
+            ]
+        rom_values = [RotationTree.angle_to_rom_value(a, self.phase_bitsize) for a in angles]
+        return rom_values
+
+
+@bloq_example
+def _state_prep_via_rotation() -> StatePreparationViaRotations:
+    state_coefs = (
+        (-0.42677669529663675 - 0.1767766952966366j),
+        (0.17677669529663664 - 0.4267766952966367j),
+        (0.17677669529663675 - 0.1767766952966368j),
+        (0.07322330470336305 - 0.07322330470336309j),
+        (0.4267766952966366 - 0.17677669529663692j),
+        (0.42677669529663664 + 0.17677669529663675j),
+        (0.0732233047033631 + 0.17677669529663678j),
+        (-0.07322330470336308 - 0.17677669529663678j),
+    )
+    state_prep_via_rotation = StatePreparationViaRotations(
+        state_bitsize=2, phase_bitsize=4, state_coefficients=state_coefs
+    )
+    return state_prep_via_rotation
+
+
+_CONTROLLED_STATE_PREP_DOC = BloqDocSpec(
+    bloq_cls=StatePreparationViaRotations,
+    import_line='from qualtran.bloqs.state_preparation.state_preparation_via_rotation import StatePreparationViaRotations',
+    examples=(_state_prep_via_rotation,),
+)
+
+
+@attrs.frozen
+class ControlledQROMRotateQubit(Bloq):
+    r"""Array of controlled rotations $Z^{\theta_i/2}$ for a list of angles $\theta$.
+
+    It uses phase kickback and thus needs a phase gradient state in order to work. This
+    state must be provided externally for efficiency, as it is unaffected and can thus be reused.
+    Refer to [1], section on arbitrary quantum state preparation on page 3.
+
+    Args:
+        selection_bitsizes: number of qubits used for encoding the selection register of the QROM, it
+            must be equal to $\lceil \log_2(l_{rv}) \rceil$, where $l_{rv}$ is the number of angles
+            provided.
+        rot_reg_bitsizes: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        rom_values: the tuple of values to be loaded in the rom, which correspond to the angle of
+            each rotation. In order to get the rom value that corresponds to a given angle use
+            RotationTree.angle_to_rom_value(angle, rot_reg_bitsizes).
+        control_bitsize: number of qubits of the control register. Set to zero for an uncontrolled gate.
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    selection_bitsizes: int
+    rot_reg_bitsizes: int
+    rom_values: Tuple[int, ...]
+    control_bitsize: int
+
+    @property
+    def signature(self) -> Signature:
+        return Signature.build(
+            prepare_control=self.control_bitsize,
+            selection=self.selection_bitsizes,
+            qubit=1,
```

Comment:
I think we should the `qubit` register with the `control_bitsize` argument so that we combine the `prepare_control` and `qubit` into a single register before calling this Bloq. It's not clear to me from the interface of this Bloq that why do we need a separate control qubit ?

The goal of this bloq is to
1) Load N multiplexed rotation angles via QROM in a selection register

$$
|x\rangle|0\rangle \rightarrow |x\rangle|\theta_x\rangle
$$

2) Apply multiplexed rotations on the target register by the loaded angle, i.e. 

$$
|x\rangle|\theta_{x}\rangle \rightarrow e^{i \pi \theta_x / 2} |x\rangle |\theta_{x}\rangle
$$

3) Uncompute step-1 s.t. you get

$$
e^{i \pi \theta_x / 2} |x\rangle |\theta_{x}\rangle \rightarrow e^{i \pi \theta_x / 2} |x\rangle |0\rangle
$$


If users specify a `control_bitsize` then steps 1 & 3 can be made controlled (via controlled `QROM`) and therefore step-2 becomes controlled for free. 

There's no reason for us to put `qubit` as part of the signature of this Bloq since it's a problem specific requirement that we should always control the i'th step of our controlled rotations using `i-1`'th qubit. 

Thus, I propose the following changes:

1) Update the name of the class to something like `ProgrammableRotationGateArrayViaPhaseGradient` or `PRGAViaPhaseGradient` or `MultiplexedRotationsViaPhaseGradient` 
2) Change the signature of the class to 
```python
Signature.build(
    control=self.control_bitsize, 
    selection=self.selection_bitsize,
    phase_grad=self.rot_reg_bitsizes,
)
```
3. Note that my suggestion is to remove the `rot_reg` from the signature of the class. This is because the `rot_reg` is a temporary ancilla register that the class needs to allocate for loading QROM angles in order to achieve the required phasing. The `rot_reg` allocation and deallocation should happen inside the `build_composite_bloq` of this class.  The signature of the class then matches the 3 step description defined above.

Reply:
Done. This add some more code that handles some special cases when joining the qubit with the control register, but I think that's unavoidable.